src/Compilers/CSharp/Portable/Symbols/DelegateCaches/DelegateCacheManager.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using Microsoft.Cci;
+using Microsoft.CodeAnalysis.CSharp.Emit;
+using Microsoft.CodeAnalysis.Symbols;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    using ModuleScopedContainerCollection = ConcurrentDictionary<TypeSymbol, ModuleScopedDelegateCacheContainer>;
+
+    internal sealed class DelegateCacheManager : IComparer<ModuleScopedDelegateCacheContainer>
+    {
+        private ModuleScopedContainerCollection _lazyModuleScopedContainers;
+
+        private ModuleScopedContainerCollection ModuleScopedContainers
+        {
+            get
+            {
+                if (_lazyModuleScopedContainers == null)
+                {
+                    Interlocked.CompareExchange(ref _lazyModuleScopedContainers, new ModuleScopedContainerCollection(), null);
+                }
+
+                return _lazyModuleScopedContainers;
+            }
+        }
+
+        internal DelegateCacheContainer ObtainModuleScopedContainer(CSharpCompilation compilation, TypeSymbol delegateType)
+        {
+            return ModuleScopedContainers.GetOrAdd(delegateType, t => new ModuleScopedDelegateCacheContainer(compilation, t));
+        }
+
+        internal void AssignNames(PEModuleBuilder moduleBuilder, DiagnosticBag diagnostics)
```

Comment:
We should ensure ObtainModuleScopedContainer isn't called after AssignNames is called.